aggregator/src/datastore.rs

```diff
@@ -8734,6 +9279,1440 @@ mod tests {
+
+                    // Leader, time-interval aggregation job with old & new reports [not GC'ed].
+                    let (aggregation_job_id, batch_identifier, report_ids) =
+                        write_aggregation_artifacts::<TimeInterval>(
+                            tx,
+                            leader_time_interval_task.id(),
+                            &[
+                                OLDEST_ALLOWED_REPORT_TIMESTAMP
+                                    .sub(&Duration::from_seconds(5))
+                                    .unwrap(),
+                                OLDEST_ALLOWED_REPORT_TIMESTAMP
+                                    .add(&Duration::from_seconds(5))
+                                    .unwrap(),
+                            ],
+                        )
+                        .await;
+                    aggregation_job_ids.insert(aggregation_job_id);
+                    batch_identifiers.insert(batch_identifier.get_encoded());
+                    all_report_ids.extend(report_ids);
+
+                    // Leader, time-interval aggregation job with new reports [not GC'ed].
+                    let (aggregation_job_id, batch_identifier, report_ids) =
+                        write_aggregation_artifacts::<TimeInterval>(
+                            tx,
+                            leader_time_interval_task.id(),
+                            &[
+                                OLDEST_ALLOWED_REPORT_TIMESTAMP
+                                    .add(&Duration::from_seconds(9))
+                                    .unwrap(),
+                                OLDEST_ALLOWED_REPORT_TIMESTAMP
+                                    .add(&Duration::from_seconds(10))
+                                    .unwrap(),
+                            ],
+                        )
+                        .await;
+                    aggregation_job_ids.insert(aggregation_job_id);
+                    batch_identifiers.insert(batch_identifier.get_encoded());
+                    all_report_ids.extend(report_ids);
+
+                    // Leader, time-interval aggregation job with attached collect job [not GC'ed].
+                    let (aggregation_job_id, batch_identifier, report_ids) =
+                        write_aggregation_artifacts::<TimeInterval>(
+                            tx,
+                            leader_time_interval_task.id(),
+                            &[
+                                OLDEST_ALLOWED_REPORT_TIMESTAMP
+                                    .sub(&Duration::from_seconds(9))
+                                    .unwrap(),
+                                OLDEST_ALLOWED_REPORT_TIMESTAMP
+                                    .sub(&Duration::from_seconds(8))
+                                    .unwrap(),
+                            ],
+                        )
+                        .await;
+                    tx.put_collect_job(&CollectJob::<0, TimeInterval, dummy_vdaf::Vdaf>::new(
```

Comment:
I think you need to add some collect jobs for this task that are related to reports that are not eligible for GC, and then check that those collect jobs don't get garbage collected. That would have caught the bug fixed in eb86863.

Reply:
This task already had an un-GC'ed collect job (see the "Leader, time-interval aggregation job with attached collect job [not GC'ed]" case).

The problem was that the batch aggregations were being GC'ed anyway due to this clause, because the batch identifiers (i.e. intervals) were identical between the aggregation job & the batch aggregation:

```
                        OR batch_aggregations.batch_identifier IN (
                            SELECT batch_identifier FROM aggregation_jobs
                            WHERE aggregation_jobs.id IN (SELECT id FROM aggregation_jobs_to_delete)
                        )
```

I added a little code to "shorten" the interval used in the batch aggregation, and verified that this change detects regression.

Reply:
This task already had an un-GC'ed collect job (see the "Leader, time-interval aggregation job with attached collect job [not GC'ed]" case).

The problem was that the batch aggregations were being GC'ed anyway due to this clause, because the batch identifiers (i.e. intervals) were identical between the aggregation job & the batch aggregation:

```
                        OR batch_aggregations.batch_identifier IN (
                            SELECT batch_identifier FROM aggregation_jobs
                            WHERE aggregation_jobs.id IN (SELECT id FROM aggregation_jobs_to_delete)
                        )
```

I added a little code to "shorten" the interval used in the batch aggregation, and verified that this change detects regression.